### ISSUE TITLE
Weird Behaviour on @extend_schema with custom schema in view

### ISSUE THREAD

**Issue by ChrisMaxheart** [tfranzel/drf-spectacular#1034]
Not sure if it is related to https://github.com/tfranzel/drf-spectacular/issues/241

To reproduce:
```python
class MyDefaultAutoSchema(AutoSchema):
    def _get_response_for_code(self, *args, **kwargs):
        default_response: dict = super()._get_response_for_code(*args, **kwargs)
        json_schema: Optional[dict] = (
            default_response.get("content", {}).get("application/json", {}).get("schema")
        )
        updated_schema: dict = {
            "type": "object",
            "properties": {
                "default_data": {"properties": {"items": json_schema}},
                "default_code": {
                    "type": "integer",
                    "default": 12345,
                },
            },
        }
        to_return: dict = deepcopy(default_response)
        to_return["content"]["application/json"]["schema"] = updated_schema
        return to_return


class MyCustomAutoSchema(AutoSchema):
    def _get_response_for_code(self, *args, **kwargs):
        default_response: dict = super()._get_response_for_code(*args, **kwargs)
        json_schema: Optional[dict] = (
            default_response.get("content", {}).get("application/json", {}).get("schema")
        )
        updated_schema: dict = {
            "type": "object",
            "properties": {
                "custom_data": {"properties": {"items": json_schema}},
                "custom_code": {
                    "type": "integer",
                    "default": 6789,
                },
            },
        }
        to_return: dict = deepcopy(default_response)
        to_return["content"]["application/json"]["schema"] = updated_schema
        return to_return

# settings
REST_FRAMEWORK = {
    "DEFAULT_SCHEMA_CLASS": "MyDefaultAutoSchema"
}

# views
class MyView(APIView):
    schema = MyCustomAutoSchema()

    @extend_schema(tags=["custom"])
    def get(self, *args, **kwargs):
        ...

```

Expected doc behaviour:
```
{"custom_code": ..., "custom_data": ...}
```

Actual doc behaviour:
```
{"default_code": .., "default_data": {"custom_code": ..., "custom_data": ...}}
```

without the @extend_schema, the behaviour works as expected, but with the @extend_schema, seems this kind of weird behaviour is happening

May I check if this is an actual issue or am I doing something wrong when updating the response body?
Thanks!

**Comment by tfranzel** [comment]
tldr: 

Do not create a diamond inheritance problem on `AutoSchema`. You violated an undocumented assumption that custom (view) schemas must be sub-classes of the default schema. 

`AutoSchema` -> `MyDefaultAutoSchema` (default) -> `MyCustomAutoSchema` (view) -> (implicit) `ExtendedSchema` (by decoration) 

This will happen regardless of how you create your classes, which is where the weird behavior is coming from.

This will fix your issue:
```python
    class MyCustomAutoSchema(MyDefaultAutoSchema):
        def _get_response_for_code(self, *args, **kwargs):
            default_response: dict = super(MyDefaultAutoSchema, self)._get_response_for_code(*args, **kwargs)
            ...
```

---

long version:

Every time you use `@extend_schema`, a modified schema subclass is created based from either the default class or a class from a previous decoration. It is essentially a layer cake and you can arbitrarily stack decorators and python class mechanics will do the heavy lifting.

However, since `MyCustomAutoSchema` is not yet available when we process `@extend_schema(tags=["custom"])` (at class creation time), we need to do some MRO trickery later. We re-arrange the class hierarchy afterwards to remedy that chicken-egg problem. It is complicated.

https://github.com/tfranzel/drf-spectacular/blob/388da8dedb3d3ec9d687146934bc6e41ef83dcf5/drf_spectacular/generators.py#L159

Not sure if anybody ever discovered this, since users usually just subclass their default class and patch one or two methods on top. Congratulations on peeking behind the curtain.

**Comment by ChrisMaxheart** [comment]
Understand, thanks for the workaround!
